File: java/QuickSort.java
package com.htl.algorithms;

import java.util.Arrays;

public class QuickSort {

	public static void main(String[] args) {

		int[] arr = { 6, 1, 7, 3, 6, 9, 0 }; // unsorted array
		         // { 0, 1, 2, 3, 4, 5, 6 } index
		int arrLength = arr.length; // array length 7
		System.out.println(arrLength);

		// call quickSort method
		quickSort(arr, 0, arrLength-1);

		for (int i : arr) {
			System.out.print(i + " ");
		}
	}

	static void swap(int[] arr, int index1, int index2) {
		int temp;
		temp = arr[index1];
		arr[index1] = arr[index2];
		arr[index2] = temp;
	}

	static int partition(int[] arr, int start, int end) {
		// taking the right most element as pivot
		int pivot = arr[end];
		int partitionIndex = start; // setting the partition index as start initially

		for (int i = start; i < end; i++) {
			if (arr[i] <= pivot) {
				swap(arr, i, partitionIndex);
				partitionIndex++;
			}
		}
		swap(arr, partitionIndex, end);
		return partitionIndex;
	}

	static void quickSort(int[] arr, int start, int end) {
		if (start < end) {
			int partitionIndex = partition(arr, start, end);
			quickSort(arr, start, partitionIndex - 1);
			quickSort(arr, partitionIndex + 1, end);
		}
	}
}

File: python/quicksort.py
# Quick sorts time complexity mostly comes out to be O(NlogN)
# Using random partition the time complexity can be reduced

def swap(arr, idx1, idx2):
	temp = arr[idx1]
	arr[idx1] = arr[idx2]
	arr[idx2] = temp

def partition(arr, start, end):
	# Choosing last element as pivot
	pivot = arr[end]
	partition_index = start
	
	for i in range(start,end):
		if arr[i] <= pivot:
			swap(arr,i,partition_index)
			partition_index = partition_index+1
	swap(arr,partition_index,end)
	return partition_index
		
def quick_sort(arr,start,end):
	if start<end:
		partition_idx = partition(arr,start,end)
		quick_sort(arr,start, partition_idx-1)
		quick_sort(arr,partition_idx+1,end)
	

arr = [6,1,7,3,6,9,0] # unsorted array
arr_length = len(arr) # array length

print(arr)

quick_sort(arr,0,arr_length-1)

print(arr)

			